# Fuel Economy Data &mdash; Part 1

### This notebook contains my initial data exploration and the data cleaning I did to generate my final dataset.

In [1]:
import pandas as pd
import numpy as np

In [2]:
# load EPA fuel economy dataset downloaded from Kaggle
df = pd.read_csv('./data/database.csv')

C:\Users\chloe\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3071: DtypeWarning: Columns (7,16,62,64,65,67,68,78) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


In [18]:
df.describe()

Vehicle ID          Year  Engine Index  Engine Cylinders  \
count  38113.000000  38113.000000  38113.000000      37977.000000   
mean   19170.638496   2000.194527   8799.389001          5.736656   
std    11134.878665     10.464573  17781.058490          1.752254   
min        1.000000   1984.000000      0.000000          2.000000   
25%     9529.000000   1991.000000      0.000000          4.000000   
50%    19058.000000   2001.000000    212.000000          6.000000   
75%    28779.000000   2009.000000   4451.000000          6.000000   
max    38542.000000   2017.000000  69102.000000         16.000000   

       Engine Displacement  City MPG (FT1)  Unrounded City MPG (FT1)  \
count         37979.000000    38113.000000              38113.000000   
mean              3.317583       17.981109                  4.606426   
std               1.361995        6.849728                 10.113963   
min               0.000000        6.000000                  0.000000   
25%               2.200000       15.000000                  0.000000   
50%               3.000000       17.000000                  0.000000   
75%               4.300000       20.000000                  0.000000   
max               8.400000      150.000000                150.000000   

       City MPG (FT2)  Unrounded City MPG (FT2)  \
count    38113.000000              38113.000000   
mean         0.546218                  0.386633   
std          4.109282                  3.861179   
min          0.000000                  0.000000   
25%          0.000000                  0.000000   
50%          0.000000                  0.000000   
75%          0.000000                  0.000000   
max        145.000000                145.083500   

       City Gasoline Consumption (CD)  ...  Hours to Charge (240V)  \
count                    38113.000000  ...            38113.000000   
mean                         0.000432  ...                0.027635   
std                          0.039976  ...                0.473685   
min                          0.000000  ...                0.000000   
25%                          0.000000  ...                0.000000   
50%                          0.000000  ...                0.000000   
75%                          0.000000  ...                0.000000   
max                          5.350000  ...               12.000000   

       Hours to Charge (AC 240V)  Composite City MPG  Composite Highway MPG  \
count               38113.000000        38113.000000           38113.000000   
mean                    0.005549            0.082203               0.080891   
std                     0.161014            2.156682               2.052187   
min                     0.000000            0.000000               0.000000   
25%                     0.000000            0.000000               0.000000   
50%                     0.000000            0.000000               0.000000   
75%                     0.000000            0.000000               0.000000   
max                     7.000000           97.000000              81.000000   

       Composite Combined MPG   Range (FT1)  City Range (FT1)  \
count            38113.000000  38113.000000      38113.000000   
mean                 0.081311      0.469708          0.426249   
std                  2.097794      9.352069          9.104702   
min                  0.000000      0.000000          0.000000   
25%                  0.000000      0.000000          0.000000   
50%                  0.000000      0.000000          0.000000   
75%                  0.000000      0.000000          0.000000   
max                 88.000000    315.000000        305.900000   

       Highway Range (FT1)  City Range (FT2)  Highway Range (FT2)  
count         38113.000000      38113.000000         38113.000000  
mean              0.419197          0.043973             0.040051  
std               9.315914          1.311628             1.169281  
min               0.000000          0.000000             0.000000  
25%      

In [17]:
df[df.isnull().any(axis=1)]

Vehicle ID  Year        Make               Model  \
0           26587  1984  Alfa Romeo           GT V6 2.5   
1           27705  1984  Alfa Romeo           GT V6 2.5   
2           26561  1984  Alfa Romeo  Spider Veloce 2000   
3           27681  1984  Alfa Romeo  Spider Veloce 2000   
4           27550  1984  AM General   DJ Po Vehicle 2WD   
...           ...   ...         ...                 ...   
38108       37564  2017       Volvo            XC60 FWD   
38109       37547  2017       Volvo            XC90 AWD   
38110       37548  2017       Volvo            XC90 AWD   
38111       37703  2017       Volvo       XC90 AWD PHEV   
38112       37687  2017       Volvo            XC90 FWD   

                                    Class              Drive  \
0                        Minicompact Cars                NaN   
1                        Minicompact Cars                NaN   
2                             Two Seaters                NaN   
3                             Two Seaters                NaN   
4             Special Purpose Vehicle 2WD      2-Wheel Drive   
...                                   ...                ...   
38108     Small Sport Utility Vehicle 2WD  Front-Wheel Drive   
38109  Standard Sport Utility Vehicle 4WD    All-Wheel Drive   
38110  Standard Sport Utility Vehicle 4WD    All-Wheel Drive   
38111  Standard Sport Utility Vehicle 4WD    All-Wheel Drive   
38112  Standard Sport Utility Vehicle 2WD  Front-Wheel Drive   

            Transmission Transmission Descriptor  Engine Index  \
0         Manual 5-Speed                     NaN          9001   
1         Manual 5-Speed                     NaN          9005   
2         Manual 5-Speed                     NaN          9002   
3         Manual 5-Speed                     NaN          9006   
4      Automatic 3-Speed                     NaN          1830   
...                  ...                     ...           ...   
38108     Automatic (S8)                     NaN            90   
38109     Automatic (S8)                     NaN            52   
38110     Automatic (S8)                     NaN            53   
38111     Automatic (S8)                     NaN            54   
38112     Automatic (S8)                     NaN            50   

      Engine Descriptor  ...  Hours to Charge (AC 240V)  Composite City MPG  \
0                 (FFS)  ...                        0.0                   0   
1        (FFS) CA model  ...                        0.0                   0   
2                 (FFS)  ...                        0.0                   0   
3        (FFS) CA model  ...                        0.0                   0   
4                 (FFS)  ...                        0.0                   0   
...                 ...  ...                        ...                 ...   
38108              SIDI  ...                        0.0                   0   
38109              SIDI  ...                        0.0                   0   
38110              SIDI  ...                        0.0                   0   
38111        SIDI; PHEV  ...                        0.0                  29   
38112              SIDI  ...                        0.0                   0   

      Composite Highway MPG Composite Combined MPG Range (FT1)  \
0                         0                      0           0   
1                         0                      0           0   
2                         0                      0           0   
3                         0                      0           0   
4                         0                      0           0   
...                     ...                    ...         ...   
38108                     0                      0           0   
38109                     0                      0           0   
38110                     0                      0           0   
38111                    32                     30           0   
38112                     0                      0           

There are many null values in this dataset. I will examine the dataset column by column to see what reasonable values for the different nulls should be.

In [58]:
def print_nulls(df):
    for column in df.columns:
        print(f"{column} Nulls: {df[df[column].isnull()].shape[0]}")
        print(f"{column} Valids: {df[df[column].notnull()].shape[0]}")

def nulls_to_df(df):
    """Returns a dataframe with a count of null and valid entries for each column in the
    original dataset."""
    d = dict()
    for column in df.columns:
        l = []
        l.append(df[df[column].isnull()].shape[0])
        l.append(df[df[column].notnull()].shape[0])
        d[column] = l
    
    new_df = pd.DataFrame.from_dict(d, orient='index', columns=['Nulls', 'Valids'])
    return new_df

In [49]:
print_nulls(df)

Vehicle ID Nulls: 0
Vehicle ID Valids: 38113
Year Nulls: 0
Year Valids: 38113
Make Nulls: 0
Make Valids: 38113
Model Nulls: 0
Model Valids: 38113
Class Nulls: 0
Class Valids: 38113
Drive Nulls: 1189
Drive Valids: 36924
Transmission Nulls: 11
Transmission Valids: 38102
Transmission Descriptor Nulls: 23068
Transmission Descriptor Valids: 15045
Engine Index Nulls: 0
Engine Index Valids: 38113
Engine Descriptor Nulls: 15420
Engine Descriptor Valids: 22693
Engine Cylinders Nulls: 136
Engine Cylinders Valids: 37977
Engine Displacement Nulls: 134
Engine Displacement Valids: 37979
Turbocharger Nulls: 32874
Turbocharger Valids: 5239
Supercharger Nulls: 37420
Supercharger Valids: 693
Fuel Type Nulls: 0
Fuel Type Valids: 38113
Fuel Type 1 Nulls: 0
Fuel Type 1 Valids: 38113
Fuel Type 2 Nulls: 36675
Fuel Type 2 Valids: 1438
City MPG (FT1) Nulls: 0
City MPG (FT1) Valids: 38113
Unrounded City MPG (FT1) Nulls: 0
Unrounded City MPG (FT1) Valids: 38113
City MPG (FT2) Nulls: 0
City MPG (FT2) Valids: 3811

In [64]:
nulls = nulls_to_df(df)
nulls.sort_values(by='Nulls', ascending=False)

Nulls  Valids
Alternate Charger                38067      46
Vehicle Charger                  38065      48
Electric Motor                   37509     604
Supercharger                     37420     693
Range (FT2)                      36680    1433
...                                ...     ...
Unadjusted Highway MPG (FT1)         0   38113
Unadjusted City MPG (FT1)            0   38113
Highway Utility Factor               0   38113
Highway Electricity Consumption      0   38113
Highway Range (FT2)                  0   38113

[81 rows x 2 columns]

With 38,067 null values and only 46 valid values, **Alternate Charger** is in the worst shape, so I'll start with that. Based on the values in the column and on information found on the [Department of Energy's website](https://www.fueleconomy.gov/feg/ws/), this column seems to describe alternate chargers for electric vehicles. This does not apply to the vast majority of vehicles, so a reasonable value to substitute is "None".

In [65]:
df['Alternate Charger'].unique()

array([nan, '6.6 kW charger', 'dual charger', '80 amp dual charger',
       '3.6 kW charger'], dtype=object)

In [67]:
df['Alternate Charger'].fillna('None', inplace=True)
df['Alternate Charger'].unique()

array(['None', '6.6 kW charger', 'dual charger', '80 amp dual charger',
       '3.6 kW charger'], dtype=object)

In [69]:
nulls = nulls_to_df(df)
nulls.sort_values(by='Nulls', ascending=False)

Nulls  Valids
Vehicle Charger                  38065      48
Electric Motor                   37509     604
Supercharger                     37420     693
Range (FT2)                      36680    1433
Fuel Type 2                      36675    1438
...                                ...     ...
Unadjusted Highway MPG (FT1)         0   38113
Unadjusted City MPG (FT1)            0   38113
Highway Utility Factor               0   38113
Highway Electricity Consumption      0   38113
Highway Range (FT2)                  0   38113

[81 rows x 2 columns]

I will clean the remaining columns similarly, relying on information from the US Department of Energy as well as column values to determine appropriate substitutions for nulls.

In [70]:
df['Vehicle Charger'].unique()

array([nan, '3.6 kW charger', 'single charger', 'standard charger',
       '6.6 kW charger', '7.2 kW charger'], dtype=object)

Vehicle charger also refers only to electric vehicles. Using the same logics as with Alternate Charger, I will subtitute these values with "None" also:

In [71]:
df['Vehicle Charger'].fillna('None', inplace=True)
nulls = nulls_to_df(df)
nulls.sort_values(by='Nulls', ascending=False)

Nulls  Valids
Electric Motor                   37509     604
Supercharger                     37420     693
Range (FT2)                      36680    1433
Fuel Type 2                      36675    1438
Gas Guzzler Tax                  35817    2296
...                                ...     ...
Unadjusted Highway MPG (FT1)         0   38113
Unadjusted City MPG (FT1)            0   38113
Highway Utility Factor               0   38113
Highway Electricity Consumption      0   38113
Highway Range (FT2)                  0   38113

[81 rows x 2 columns]

In [72]:
df['Electric Motor'].unique()

array([nan, '85 kW AC Induction', '49 kW DC Brushless',
       '56kW AC Induction', '67 KW AC  Induction', '67 KW AC',
       '102kW AC Induction', '49kW DC Brushless', '62 KW AC Induction',
       '50 KW DC', '67 KW AC Induction', '27 KW AC Induction',
       '24 KW AC Synchronous', '36V Ni-MH', '288V Ni-MH', '330V Ni-MH',
       '158V Ni-MH', '150 kW', '245V Ni-MH', '202V Ni-MH', '300V Ni-MH',
       '312V Ni-MH', '275V Ni-MH', '101V Ni-MH', '126V Li-Ion',
       '125 kW AC Induction', '111 kW', '270V Li-Ion', '80 kW DCPM',
       '30 kW DCPM', '52 kW AC Induction', '374V Li-Ion', '115V Li-Ion',
       '75 kW AC PMSM', '100 kW DCPM', '2 @ 150 kw (300 kw)',
       '107 kW AC Induction', '144V Li-Ion', '346V Li-Ion', '49 kW DCPM',
       '260 kW AC Induction', '144V Ni-MH', '18 kW',
       '115 kW AC Induction', '266V Li-Ion', '82 kW ACIPM', '68 kW',
       '280V Li-Ion', '107 kW AC PMSM', '92 kW DCPM', '55 kW DCPM',
       '225 kW AC Induction', '270 kW AC Induction', '220V Li-Ion',
 

The **Electric Motor** column describes the kilowatt hours for the vehicle's electric motor. This doesn't apply to most vehicles, so substituting "None" is appropriate.

In [73]:
df['Electric Motor'].fillna('None', inplace=True)
nulls = nulls_to_df(df)
nulls.sort_values(by='Nulls', ascending=False)

Nulls  Valids
Supercharger                     37420     693
Range (FT2)                      36680    1433
Fuel Type 2                      36675    1438
Gas Guzzler Tax                  35817    2296
Alternative Fuel/Technology      35066    3047
...                                ...     ...
Unadjusted Highway MPG (FT1)         0   38113
Unadjusted City MPG (FT1)            0   38113
Highway Utility Factor               0   38113
Highway Electricity Consumption      0   38113
Highway Range (FT2)                  0   38113

[81 rows x 2 columns]

In [74]:
df['Supercharger'].unique()

array([nan, 'S'], dtype=object)

The **Supercharger** column is different from what we've seen so far. It contains an "S" if the vehicle is supercharged and nothing otherwise. I think substituting "NS" (denoting "not supercharged") is a reasonable thing to do here:

In [75]:
df['Supercharger'].fillna('NS', inplace=True)
nulls = nulls_to_df(df)
nulls.sort_values(by='Nulls', ascending=False)

Nulls  Valids
Range (FT2)                        36680    1433
Fuel Type 2                        36675    1438
Gas Guzzler Tax                    35817    2296
Alternative Fuel/Technology        35066    3047
Turbocharger                       32874    5239
...                                  ...     ...
Unadjusted City MPG (FT1)              0   38113
Highway Utility Factor                 0   38113
Highway Electricity Consumption        0   38113
Highway Gasoline Consumption (CD)      0   38113
Highway Range (FT2)                    0   38113

[81 rows x 2 columns]

In [77]:
df['Range (FT2)'].unique()

array([nan, '360', '300', '230/270/270', '240/290/290', '220/270/260',
       '240/290/280', '290', '230/270', '240/280', '220/260', '140',
       '310', '280', '230', '210', '260', '120', '250/220/320', '270',
       '110', '290/340', '320', '310/390', '280/380', '130', '180',
       '270/240/340', '170', '310-540', '310/460', '310/540', '310/330',
       '310/480', '310/440', '330', '340-400', '370', '340', '340-440',
       '310-410', '240', '340-350', '290/410', '310/340', '250/370',
       '230/350', '240-420', '250', '310/320', '280/370', '350-370',
       '280-350', '310-420', '310-340', '330-450', '300-420', '310/410',
       '380', '410', '350', '310/420', '330/450', '220', '260/320',
       '310/370', '214', '236', '279', '312', '333', '307', '304', '383',
       '353', '413', '295', '294', '344', '338', '277', '308', '264',
       '303', '307/420', '337/451', '347', '286/382', '329', '35', '262',
       '292', '219', '354', '265', '366', '331', '245', '282', '337',
       '6

In [96]:
df[df['Range (FT2)'].notnull()][['Range (FT2)', 
                                 'City Range (FT2)', 'Highway Range (FT2)']].sort_values(
    by='Range (FT2)', ascending=False)

Range (FT2)  City Range (FT2)  Highway Range (FT2)
37015          97           103.030               90.550
34522          72            77.504               65.632
35769          72            77.504               65.632
33304          72            77.504               65.632
30089          69             0.000                0.000
...           ...               ...                  ...
33078          11            12.150                9.100
31911          11            12.150                9.100
35539          11            12.150                9.100
36491          11            11.370               11.190
34284          11            12.150                9.100

[1433 rows x 3 columns]

In [97]:
df.dtypes

Vehicle ID               int64
Year                     int64
Make                    object
Model                   object
Class                   object
                        ...   
City Range (FT1)       float64
Highway Range (FT1)    float64
Range (FT2)             object
City Range (FT2)       float64
Highway Range (FT2)    float64
Length: 81, dtype: object

**Range (FT2)** is trickier. The column denotes the range of vehicles using their second fuel type (FT2 = fuel type 2). The integer-like values of this column are an average of the City Range (FT2) and Highway Range (FT2) columns. However, inspecting the datatypes of this dataframe, I can see that there are also some string values in this column. I need to find these, make sense of them, and, if necessary, convert them to integers. 

In [162]:
df[df['Range (FT2)'].apply(type) == str]

Vehicle ID  Year       Make                                    Model  \
18087       16403  2000  Chevrolet                   Cavalier (Bi-fuel CNG)   
18162       16422  2000   Chrysler  Town and Country/Voyager/Grand Voy. 2WD   
18189       16421  2000      Dodge                Caravan/Grand Caravan 2WD   
18291       16411  2000       Ford                           Ranger 2WD FFV   
18292       16412  2000       Ford                           Ranger 2WD FFV   
...           ...   ...        ...                                      ...   
38033       38531  2017     Toyota                              Prius Prime   
38042       38291  2017     Toyota                          Sequoia 4WD FFV   
38054       38122  2017     Toyota                           Tundra 2WD FFV   
38057       38127  2017     Toyota                           Tundra 4WD FFV   
38111       37703  2017      Volvo                            XC90 AWD PHEV   

                                    Class                    Drive  \
18087                     Subcompact Cars        Front-Wheel Drive   
18162                       Minivan - 2WD        Front-Wheel Drive   
18189                       Minivan - 2WD        Front-Wheel Drive   
18291          Standard Pickup Trucks 2WD         Rear-Wheel Drive   
18292          Standard Pickup Trucks 2WD         Rear-Wheel Drive   
...                                   ...                      ...   
38033                        Midsize Cars        Front-Wheel Drive   
38042  Standard Sport Utility Vehicle 4WD  Part-time 4-Wheel Drive   
38054          Standard Pickup Trucks 2WD         Rear-Wheel Drive   
38057          Standard Pickup Trucks 4WD  Part-time 4-Wheel Drive   
38111  Standard Sport Utility Vehicle 4WD          All-Wheel Drive   

                           Transmission Transmission Descriptor  Engine Index  \
18087                 Automatic 3-Speed                     NaN             0   
18162                 Automatic 4-Speed                   CLKUP             0   
18189                 Automatic 4-Speed                   CLKUP             0   
18291                 Automatic 4-Speed                   CLKUP             0   
18292                    Manual 5-Speed                     NaN             0   
...                                 ...                     ...           ...   
38033  Automatic (variable gear ratios)                     NaN           108   
38042                    Automatic (S6)                     NaN             8   
38054                    Automatic (S6)                     NaN             9   
38057                    Automatic (S6)                     NaN            10   
38111                    Automatic (S8)                     NaN            54   

      Engine Descriptor  ...  Hours to Charge (AC 240V)  Composite City MPG  \
18087              NONE  ...                        0.0                   0   
18162         FLEX-FUEL  ...                        0.0                   0   
18189         FLEX-FUEL  ...                        0.0                   0   
18291         FLEX-FUEL  ...                        0.0                   0   
18292         FLEX-FUEL  ...                        0.0                   0   
...                 ...  ...                        ...                 ...   
38033              PHEV  ...                        0.0                  83   
38042               FFV  ...                        0.0                   0   
38054               FFV  ...                        0.0                   0   
38057               FFV  ...                        0.0                   0   
38111        SIDI; PHEV  ...                        0.0                  29   

      Composite Highway MPG Composite Combined MPG Range (FT1)  \
18087                     0                      0           0   
18162                     0                      0           0   
18189                     0                      0           0   
18291                     0    

In [182]:
df[df['Range (FT2)'].apply(lambda x: len(str(x)) > 3)]

Vehicle ID  Year    Make           Model                       Class  \
18291       16411  2000    Ford  Ranger 2WD FFV  Standard Pickup Trucks 2WD   
18292       16412  2000    Ford  Ranger 2WD FFV  Standard Pickup Trucks 2WD   
18293       16418  2000    Ford  Ranger 4WD FFV  Standard Pickup Trucks 4WD   
18294       16419  2000    Ford  Ranger 4WD FFV  Standard Pickup Trucks 4WD   
18477       16413  2000   Mazda   B3000 FFV 2WD  Standard Pickup Trucks 2WD   
...           ...   ...     ...             ...                         ...   
37378       37660  2017     GMC  Sierra K15 4WD  Standard Pickup Trucks 4WD   
37953       37992  2017     Ram        1500 2WD  Standard Pickup Trucks 2WD   
37956       37995  2017     Ram        1500 4WD  Standard Pickup Trucks 4WD   
38054       38122  2017  Toyota  Tundra 2WD FFV  Standard Pickup Trucks 2WD   
38057       38127  2017  Toyota  Tundra 4WD FFV  Standard Pickup Trucks 4WD   

                            Drive       Transmission Transmission Descriptor  \
18291            Rear-Wheel Drive  Automatic 4-Speed                   CLKUP   
18292            Rear-Wheel Drive     Manual 5-Speed                     NaN   
18293  4-Wheel or All-Wheel Drive  Automatic 4-Speed                   CLKUP   
18294  4-Wheel or All-Wheel Drive     Manual 5-Speed                     NaN   
18477            Rear-Wheel Drive  Automatic 4-Speed                   CLKUP   
...                           ...                ...                     ...   
37378               4-Wheel Drive  Automatic 6-Speed                     NaN   
37953            Rear-Wheel Drive  Automatic 8-Speed                     NaN   
37956               4-Wheel Drive  Automatic 8-Speed                     NaN   
38054            Rear-Wheel Drive     Automatic (S6)                     NaN   
38057     Part-time 4-Wheel Drive     Automatic (S6)                     NaN   

       Engine Index Engine Descriptor  ...  Hours to Charge (AC 240V)  \
18291             0         FLEX-FUEL  ...                        0.0   
18292             0         FLEX-FUEL  ...                        0.0   
18293             0         FLEX-FUEL  ...                        0.0   
18294             0         FLEX-FUEL  ...                        0.0   
18477             0         FLEX-FUEL  ...                        0.0   
...             ...               ...  ...                        ...   
37378           639         SIDI; FFV  ...                        0.0   
37953           681               FFV  ...                        0.0   
37956           690               FFV  ...                        0.0   
38054             9               FFV  ...                        0.0   
38057            10               FFV  ...                        0.0   

       Composite City MPG Composite Highway MPG Composite Combined MPG  \
18291                   0                     0                      0   
18292                   0                     0                      0   
18293                   0                     0                      0   
18294                   0                     0                      0   
18477                   0                     0                      0   
...                   ...                   ...                    ...   
37378                   0                     0                      0   
37953                   0                     0                      0   
37956                   0                     0                      0   
38054                   0                     0                      0   
38057                   0                     0                      0   

      Range (FT1) City Range (FT1) Highway Range (FT1)  Range (FT2)  \
18291           0              0.0                 0.0  230/270/270   
18292           0              0.0                 0.0  240/290/290   
18293           0              0.0                 0.0  220/270/260   
18294           0              0.0             

Additional research online revealed that the "slash" notation (e.g. 230/270/270) in the Range (FT2) column denotes differences between pickup trucks with "regular cab short wheelbase, regular cab long wheelbase, and super cab models, respectively". I found this information in a report on [FuelEconomy.gov](https://www.fueleconomy.gov/FEG/pdfs/guides/FEG2000_Part1.pdf).

To simplify this column, I think it's reasonable to take the average of the values and use that single value instead. First, though, I will replace null values with 0.

In [185]:
df['Range (FT2)'].fillna(0, inplace=True)

In [228]:
def convert_rangeft(series):
    """Returns a version of the Range (FT#) column with only integers and 
    uses the mean for rows with multiple values listed."""
    l = []
    for i in series:
        try:
            l.append(int(i))
        except:
            if '/' in i:
                i = i.split('/')
            elif '-' in i:
                i = i.split('-')
            i = [int(num) for num in i]
            avg = round(sum(i)/len(i))
            l.append(avg)
    return pd.Series(l)
            
                

In [229]:
df['Range (FT2) Clean'] = convert_rangeft(df['Range (FT2)'])

In [230]:
df[df['Range (FT2) Clean'] > 0]

Vehicle ID  Year       Make                                    Model  \
18087       16403  2000  Chevrolet                   Cavalier (Bi-fuel CNG)   
18162       16422  2000   Chrysler  Town and Country/Voyager/Grand Voy. 2WD   
18189       16421  2000      Dodge                Caravan/Grand Caravan 2WD   
18291       16411  2000       Ford                           Ranger 2WD FFV   
18292       16412  2000       Ford                           Ranger 2WD FFV   
...           ...   ...        ...                                      ...   
38033       38531  2017     Toyota                              Prius Prime   
38042       38291  2017     Toyota                          Sequoia 4WD FFV   
38054       38122  2017     Toyota                           Tundra 2WD FFV   
38057       38127  2017     Toyota                           Tundra 4WD FFV   
38111       37703  2017      Volvo                            XC90 AWD PHEV   

                                    Class                    Drive  \
18087                     Subcompact Cars        Front-Wheel Drive   
18162                       Minivan - 2WD        Front-Wheel Drive   
18189                       Minivan - 2WD        Front-Wheel Drive   
18291          Standard Pickup Trucks 2WD         Rear-Wheel Drive   
18292          Standard Pickup Trucks 2WD         Rear-Wheel Drive   
...                                   ...                      ...   
38033                        Midsize Cars        Front-Wheel Drive   
38042  Standard Sport Utility Vehicle 4WD  Part-time 4-Wheel Drive   
38054          Standard Pickup Trucks 2WD         Rear-Wheel Drive   
38057          Standard Pickup Trucks 4WD  Part-time 4-Wheel Drive   
38111  Standard Sport Utility Vehicle 4WD          All-Wheel Drive   

                           Transmission Transmission Descriptor  Engine Index  \
18087                 Automatic 3-Speed                     NaN             0   
18162                 Automatic 4-Speed                   CLKUP             0   
18189                 Automatic 4-Speed                   CLKUP             0   
18291                 Automatic 4-Speed                   CLKUP             0   
18292                    Manual 5-Speed                     NaN             0   
...                                 ...                     ...           ...   
38033  Automatic (variable gear ratios)                     NaN           108   
38042                    Automatic (S6)                     NaN             8   
38054                    Automatic (S6)                     NaN             9   
38057                    Automatic (S6)                     NaN            10   
38111                    Automatic (S8)                     NaN            54   

      Engine Descriptor  ...  Composite City MPG  Composite Highway MPG  \
18087              NONE  ...                   0                      0   
18162         FLEX-FUEL  ...                   0                      0   
18189         FLEX-FUEL  ...                   0                      0   
18291         FLEX-FUEL  ...                   0                      0   
18292         FLEX-FUEL  ...                   0                      0   
...                 ...  ...                 ...                    ...   
38033              PHEV  ...                  83                     73   
38042               FFV  ...                   0                      0   
38054               FFV  ...                   0                      0   
38057               FFV  ...                   0                      0   
38111        SIDI; PHEV  ...                  29                     32   

      Composite Combined MPG Range (FT1) City Range (FT1) Highway Range (FT1)  \
18087                      0           0              0.0                 0.0   
18162                      0           0              0.0                 0.0   
18189                      0           0              0.0                 0.0   
18291              

Now that I've computed Range (FT2) averages where necessary, and converted null values to zeroes, I now have a clean column that contains a single data type, int64:

In [232]:
df.dtypes

Vehicle ID               int64
Year                     int64
Make                    object
Model                   object
Class                   object
                        ...   
Highway Range (FT1)    float64
Range (FT2)             object
City Range (FT2)       float64
Highway Range (FT2)    float64
Range (FT2) Clean        int64
Length: 82, dtype: object

In [187]:
nulls = nulls_to_df(df)
nulls.sort_values(by='Nulls', ascending=False)

Nulls  Valids
Fuel Type 2                        36675    1438
Gas Guzzler Tax                    35817    2296
Alternative Fuel/Technology        35066    3047
Turbocharger                       32874    5239
Start Stop Technology              31670    6443
...                                  ...     ...
Unadjusted City MPG (FT1)              0   38113
Highway Utility Factor                 0   38113
Highway Electricity Consumption        0   38113
Highway Gasoline Consumption (CD)      0   38113
Highway Range (FT2)                    0   38113

[81 rows x 2 columns]

In [188]:
df['Fuel Type 2'].unique()

array([nan, 'Natural Gas', 'E85', 'Propane', 'Electricity'], dtype=object)

Fuel Type 2 null values can be replaced with "None". Not all vehicles will have an alternate fuel type that they can use:

In [189]:
df['Fuel Type 2'].fillna('None', inplace=True)
nulls = nulls_to_df(df)
nulls.sort_values(by='Nulls', ascending=False)

Nulls  Valids
Gas Guzzler Tax                    35817    2296
Alternative Fuel/Technology        35066    3047
Turbocharger                       32874    5239
Start Stop Technology              31670    6443
Manufacturer Code                  30794    7319
...                                  ...     ...
Unadjusted City MPG (FT1)              0   38113
Highway Utility Factor                 0   38113
Highway Electricity Consumption        0   38113
Highway Gasoline Consumption (CD)      0   38113
Highway Range (FT2)                    0   38113

[81 rows x 2 columns]

In [190]:
df['Gas Guzzler Tax'].unique()

array([nan, 'T', 'G', 'S'], dtype=object)

[FuelEconomy.gov](https://www.fueleconomy.gov/feg/ws/#emissions) states that for the **Gas Guzzler Tax** column, vehicles with "G" or "T" values are subject to the gas guzzler tax. However, the website does not mention what "S" denotes in this case. I will follow my usual convention of converting null values to "None".

In [191]:
df['Gas Guzzler Tax'].fillna('None', inplace=True)
nulls = nulls_to_df(df)
nulls.sort_values(by='Nulls', ascending=False)

Nulls  Valids
Alternative Fuel/Technology        35066    3047
Turbocharger                       32874    5239
Start Stop Technology              31670    6443
Manufacturer Code                  30794    7319
Transmission Descriptor            23068   15045
...                                  ...     ...
Unadjusted City MPG (FT1)              0   38113
Highway Utility Factor                 0   38113
Highway Electricity Consumption        0   38113
Highway Gasoline Consumption (CD)      0   38113
Highway Range (FT2)                    0   38113

[81 rows x 2 columns]

In [233]:
df['Alternative Fuel/Technology'].unique()

array([nan, 'Diesel', 'CNG', 'EV', 'Bifuel (CNG)', 'FFV', 'Hybrid',
       'Bifuel (LPG)', 'Plug-in Hybrid'], dtype=object)

According to [FuelEconomy.gov](https://www.fueleconomy.gov/feg/ws/#emissions), the **Alternative Fuel/Technology** column refers to the type of "alternative fuel or advanced technology vehicle".

- Bifuel (CNG) - Bi-fuel gasoline and compressed natural gas vehicle
- Bifuel (LPG) - Bi-fuel gasoline and propane vehicle
- CNG - Compressed natural gas vehicle
- Diesel - Diesel vehicle
- EV - Electric vehicle
- FFV - Flexible fueled vehicle (gasoline or E85)
- Hybrid - Hybrid vehicle
- Plug-in Hybrid - Plug-in hybrid vehicle

Vehicles without any values for this column do not have these features. I will replace the null values with "None".

In [234]:
df['Alternative Fuel/Technology'].fillna('None', inplace=True)
nulls = nulls_to_df(df)
nulls.sort_values(by='Nulls', ascending=False)

Nulls  Valids
Turbocharger                       32874    5239
Start Stop Technology              31670    6443
Manufacturer Code                  30794    7319
Transmission Descriptor            23068   15045
Engine Descriptor                  15420   22693
...                                  ...     ...
Unadjusted City MPG (FT1)              0   38113
Highway Utility Factor                 0   38113
Highway Electricity Consumption        0   38113
Highway Gasoline Consumption (CD)      0   38113
Range (FT2) Clean                      0   38113

[82 rows x 2 columns]

In [235]:
df['Turbocharger'].unique()

array([nan, 'T'], dtype=object)

The **Turbocharger** column is analogous to the Supercharger column above. In this case, "T" indicates that the vehicle is turbocharged. Similar to what I did for Supercharger, I will denote non-turbocharged vehicles as "NT".

In [236]:
df['Turbocharger'].fillna('NT', inplace=True)
nulls = nulls_to_df(df)
nulls.sort_values(by='Nulls', ascending=False)

Nulls  Valids
Start Stop Technology              31670    6443
Manufacturer Code                  30794    7319
Transmission Descriptor            23068   15045
Engine Descriptor                  15420   22693
Drive                               1189   36924
...                                  ...     ...
Highway Utility Factor                 0   38113
Highway Electricity Consumption        0   38113
Highway Gasoline Consumption (CD)      0   38113
Unrounded Highway MPG (FT2)            0   38113
Range (FT2) Clean                      0   38113

[82 rows x 2 columns]

In [237]:
df['Start Stop Technology'].unique()

array([nan, 'N', 'Y'], dtype=object)

In [241]:
df[df['Start Stop Technology'].isnull()]['Year'].max()

2012

According to FuelEconomy.gov, the **Start-Stop Technology** field is left blank for older vehicles. The latest model year for vehicles with null values is 2012. However, according to the [Wikipedia page](https://en.wikipedia.org/wiki/Start-stop_system#History) for this technology, there were vehicles using a version of this technology as early as 1974, so I can't just assume that all older models lacked this feature. Given that the original data treats "no" and "unknown" as categorically different, I will replace the nulls with "Unknown".

In [242]:
df['Start Stop Technology'].fillna('Unknown', inplace=True)
nulls = nulls_to_df(df)
nulls.sort_values(by='Nulls', ascending=False)

Nulls  Valids
Manufacturer Code                  30794    7319
Transmission Descriptor            23068   15045
Engine Descriptor                  15420   22693
Drive                               1189   36924
Engine Cylinders                     136   37977
...                                  ...     ...
Highway Utility Factor                 0   38113
Highway Electricity Consumption        0   38113
Highway Gasoline Consumption (CD)      0   38113
Unrounded Highway MPG (FT2)            0   38113
Range (FT2) Clean                      0   38113

[82 rows x 2 columns]

In [243]:
df['Manufacturer Code'].unique()

array([nan, 'GMX', 'HNX', 'ASX', 'BMX', 'CRX', 'FEX', 'FMX', 'JCX', 'KMX',
       'LTX', 'MBX', 'NSX', 'PRX', 'RII', 'SAX', 'TYX', 'SKR', 'ADX',
       'VWX', 'AZD', 'BEX', 'BGT', 'BYD', 'CDA', 'FSK', 'HYX', 'NLX',
       'LRX', 'MAX', 'TKX', 'MLN', 'DSX', 'MTX', 'RRG', 'FJX', 'SKX',
       'TSL', 'VVX', 'TVP', 'JLX', 'MBV', 'PGN', 'FTG', 'VGA', 'QTM'],
      dtype=object)

In [274]:
man_codes = df[df['Manufacturer Code'].notnull()][['Make','Manufacturer Code']].groupby(by='Make').describe()
man_codes.sort_values(by=('Manufacturer Code', 'unique'), ascending=False)

Manufacturer Code                 
                                  count unique  top freq
Make                                                    
Volkswagen                          210      4  VGA  108
Bentley                              46      4  VGA   24
Audi                                249      3  VGA  128
Lamborghini                          34      3  VGA   16
Chevrolet                           516      3  GMX  511
Suzuki                               37      2  SKX   33
Saab                                 27      2  SAX   21
Alfa Romeo                            6      2  CRX    5
Bugatti                               4      2  BGT    3
Land Rover                           60      2  JLX   46
Jaguar                              124      2  JLX   97
Mitsubishi                          107      2  MTX  101
Pagani                                2      1  PGN    2
Nissan                              291      1  NSX  291
Acura                                81      1  HNX   81
Mobility Ventures LLC                 2      1  MBV    2
Porsche                             339      1  PRX  339
Mercedes-Benz                       463      1  MBX  463
McLaren Automotive                   13      1  MLN   13
Pontiac                               2      1  GMX    2
SRT                                   2      1  CRX    2
Ram                                  49      1  CRX   49
Rolls-Royce                          42      1  RRG   42
Roush Performance                    20      1  RII   20
Maybach                               9      1  MBX    9
Scion                                42      1  TYX   42
Spyker                                1      1  SKR    1
Subaru                              128      1  FJX  128
Tesla                                42      1  TSL   42
Toyota                              338      1  TYX  338
VPG                                   3      1  TVP    3
Volvo                               114      1  VVX  114
Mazda                               142      1  TKX  142
Lexus                               167      1  TYX  167
Maserati                             36      1  MAX   36
Fiat                                 51      1  CRX   51
Aston Martin                         48      1  ASX   48
Azure Dynamics                        2      1  AZD    2
BMW                                 619      1  BMX  619
BYD                                   4      1  BYD    4
Buick                               102      1  GMX  102
CODA Automotive                       2      1  CDA    2
Cadillac                            174      1  GMX  174
Chrysler                             76      1  CRX   76
Dodge                               201      1  CRX  201
Ferrari                              81      1  FEX   81
Fisker                                1      1  FSK    1
MINI                                231      1  BMX  231
Ford                                513      1  FMX  513
GMC                                 289      1  GMX  289
Genesis                               7      1  HYX    7
Honda                               183      1  HNX  183
Hyundai                             241      1  HYX  241
Infiniti                            160      1  NSX  160
Jeep                                193      1  CRX  193
Kia                                 236      1  KMX  236
Lincoln                             115      1  FMX  115
Lotus                                20      1  LTX   20
smart                                22      1  MBX   22

Most makes correspond with a single **Manufacturer Code**. For makes with more than one, it makes sense to use the more frequent value. I can do this by mapping. If the list of null manufacturers codes contains makes that are not in this list, I will use "Unknown" instead.

In [317]:
top_codes = dict(man_codes[('Manufacturer Code', 'top')])
top_codes

{'Acura': 'HNX',
 'Alfa Romeo': 'CRX',
 'Aston Martin': 'ASX',
 'Audi': 'VGA',
 'Azure Dynamics': 'AZD',
 'BMW': 'BMX',
 'BYD': 'BYD',
 'Bentley': 'VGA',
 'Bugatti': 'BGT',
 'Buick': 'GMX',
 'CODA Automotive': 'CDA',
 'Cadillac': 'GMX',
 'Chevrolet': 'GMX',
 'Chrysler': 'CRX',
 'Dodge': 'CRX',
 'Ferrari': 'FEX',
 'Fiat': 'CRX',
 'Fisker': 'FSK',
 'Ford': 'FMX',
 'GMC': 'GMX',
 'Genesis': 'HYX',
 'Honda': 'HNX',
 'Hyundai': 'HYX',
 'Infiniti': 'NSX',
 'Jaguar': 'JLX',
 'Jeep': 'CRX',
 'Kia': 'KMX',
 'Lamborghini': 'VGA',
 'Land Rover': 'JLX',
 'Lexus': 'TYX',
 'Lincoln': 'FMX',
 'Lotus': 'LTX',
 'MINI': 'BMX',
 'Maserati': 'MAX',
 'Maybach': 'MBX',
 'Mazda': 'TKX',
 'McLaren Automotive': 'MLN',
 'Mercedes-Benz': 'MBX',
 'Mitsubishi': 'MTX',
 'Mobility Ventures LLC': 'MBV',
 'Nissan': 'NSX',
 'Pagani': 'PGN',
 'Pontiac': 'GMX',
 'Porsche': 'PRX',
 'Ram': 'CRX',
 'Rolls-Royce': 'RRG',
 'Roush Performance': 'RII',
 'SRT': 'CRX',
 'Saab': 'SAX',
 'Scion': 'TYX',
 'Spyker': 'SKR',
 'Subaru':

In [320]:
df['Manufacturer Code Clean'] = np.where(df['Manufacturer Code'].isnull(), df['Make'].map(top_codes) ,df['Manufacturer Code'])

In [323]:
df[df['Manufacturer Code Clean'].isnull()]

Vehicle ID  Year                         Make              Model  \
4           27550  1984                   AM General  DJ Po Vehicle 2WD   
5           28426  1984                   AM General  DJ Po Vehicle 2WD   
6           27549  1984                   AM General   FJ8c Post Office   
7           28425  1984                   AM General   FJ8c Post Office   
8           27593  1984  American Motors Corporation          Eagle 4WD   
...           ...   ...                          ...                ...   
30513       30361  2011                      Mercury          Milan FWD   
30514       30363  2011                      Mercury          Milan FWD   
30515       30375  2011                      Mercury      Milan FWD FFV   
30516       30370  2011                      Mercury   Milan Hybrid FWD   
30517       30374  2011                      Mercury        Milan S FWD   

                             Class                       Drive  \
4      Special Purpose Vehicle 2WD               2-Wheel Drive   
5      Special Purpose Vehicle 2WD               2-Wheel Drive   
6      Special Purpose Vehicle 2WD               2-Wheel Drive   
7      Special Purpose Vehicle 2WD               2-Wheel Drive   
8      Special Purpose Vehicle 4WD  4-Wheel or All-Wheel Drive   
...                            ...                         ...   
30513                 Midsize Cars           Front-Wheel Drive   
30514                 Midsize Cars           Front-Wheel Drive   
30515                 Midsize Cars           Front-Wheel Drive   
30516                 Midsize Cars           Front-Wheel Drive   
30517                 Midsize Cars           Front-Wheel Drive   

                           Transmission Transmission Descriptor  Engine Index  \
4                     Automatic 3-Speed                     NaN          1830   
5                     Automatic 3-Speed                     NaN          1880   
6                     Automatic 3-Speed                     NaN          1831   
7                     Automatic 3-Speed                     NaN          1881   
8                     Automatic 3-Speed                     NaN          1524   
...                                 ...                     ...           ...   
30513                 Automatic 6-Speed                     NaN            57   
30514                    Manual 6-Speed                     NaN            63   
30515                    Automatic (S6)                     NaN           280   
30516  Automatic (variable gear ratios)                     NaN            81   
30517                    Manual 6-Speed                     NaN           191   

      Engine Descriptor  ...  Composite Highway MPG  Composite Combined MPG  \
4                 (FFS)  ...                      0                       0   
5        (FFS) CA model  ...                      0                       0   
6                 (FFS)  ...                      0                       0   
7        (FFS) CA model  ...                      0                       0   
8                 (FFS)  ...                      0                       0   
...                 ...  ...                    ...                     ...   
30513               NaN  ...                      0                       0   
30514               NaN  ...                      0                       0   
30515               FFV  ...                      0                       0   
30516               NaN  ...                      0                       0   
30517               NaN  ...                      0                       0   

      Range (FT1) City Range (FT1) Highway Range (FT1) Range (FT2)  \
4               0              0.0                 0.0           0   
5               0              0.0                 0.0           0   
6               0              0.0                 0.0           0   
7               0              0.0                 0.0           0   
8               0              0.0       

This reduced the number of nulls from over 30,000 to around 3,000. The remaining 3,000 nulls correspond to vehicles whose make was not in the top_codes list. I will replace them with "Unknown" then drop the old Manufacturer Code column:

In [330]:
df['Manufacturer Code Clean'].fillna('Unknown', inplace=True)
df = df.drop('Manufacturer Code', axis=1)
nulls = nulls_to_df(df)
nulls.sort_values(by='Nulls', ascending=False)

Nulls  Valids
Transmission Descriptor            23068   15045
Engine Descriptor                  15420   22693
Drive                               1189   36924
Engine Cylinders                     136   37977
Engine Displacement                  134   37979
...                                  ...     ...
Highway Utility Factor                 0   38113
Highway Electricity Consumption        0   38113
Highway Gasoline Consumption (CD)      0   38113
Unrounded Highway MPG (FT2)            0   38113
Manufacturer Code Clean                0   38113

[82 rows x 2 columns]

In [331]:
df['Transmission Descriptor'].unique()

array([nan, 'Creeper', 'SIL', 'Overdrive', 'Lockup', 'VLKUP', '2MODE',
       '2MODE 2LKUP', '3LKUP', '2LKUP', 'CLKUP', '2MODE 3LKUP',
       'fuel injection', 'SIL Creeper', 'DC/FW', 'Lockup A3',
       '2MODE CLKUP', 'EMS 2MODE CLKU', 'EMS', '3MODE VLKUP', '4MODE',
       '3MODE CLKUP', 'SIL CMODE', '2MODE VLKUP', 'EMS 2MODE CLKUP',
       'EMS 5MODE', 'EMS 2MODE', '2MODE DC/FW', 'Elec Overdrive',
       'Mech Overdrive', '3MODE', 'CMODE', 'EMS 3MODE', '2MODE CLKUP FW',
       '3MODE 2LKUP', 'VMODE CLKUP', 'EMS CLKUP', 'EMS CMODE CLKUP',
       'SIL CLKUP', 'CMODE CLKUP', 'VMODE', 'EMS 3MODE CLKUP',
       '4MODE CLKUP', '6MODE', 'SIL EMS', 'VMODE VLKUP', '6MODE CLKUP',
       'SIL 3MODE', 'LONG RATIO', 'CMODE VLKUP', 'SIL 3MODE CLKUP',
       '3MODE CLKUP FW', 'SIL 2MODE CLKUP'], dtype=object)

Descriptions of Transmission Descriptors can be found on [FuelEconomy.gov](https://www.fueleconomy.gov/feg/findacarhelp.shtml#trany). Missing values can be replaced with "Unknown":

In [332]:
df['Transmission Descriptor'].fillna('Unknown', inplace=True)
nulls = nulls_to_df(df)
nulls.sort_values(by='Nulls', ascending=False)

Nulls  Valids
Engine Descriptor                  15420   22693
Drive                               1189   36924
Engine Cylinders                     136   37977
Engine Displacement                  134   37979
Transmission                          11   38102
...                                  ...     ...
Highway Utility Factor                 0   38113
Highway Electricity Consumption        0   38113
Highway Gasoline Consumption (CD)      0   38113
Unrounded Highway MPG (FT2)            0   38113
Manufacturer Code Clean                0   38113

[82 rows x 2 columns]

In [333]:
df['Engine Descriptor'].unique()

array(['(FFS)', '(FFS) CA model', '(FFS)(SIL) CA model', '(GUZZLER)',
       '(GUZZLER) CA model', '(FFS,TRBO)', '(FFS,TRBO) CA model',
       '(DIESEL)', '(DIESEL) CA model', '(CAL,FFS)', '(CAL)(FFS)',
       '(GM-OLDS)', '(350 V8) (DIESEL)', '(GM-OLDS) CA model',
       '(GM-BUICK)', '(GM-BUICK) CA model', nan, '(GM-CHEV)',
       '(GM-CHEV) CA model', '(350 V8) (FFS) CA model', '(FFS)(SIL)',
       '(350 V8) (FFS)', '(350 V8)', 'DIESEL CA model', 'CA model',
       '(CALIF)', '(DSL,TRBO)', '(DSL,TRBO) CA model', '(CALIF) CA model',
       '(FFS,MPFI) CA model', '(FFS,SPFI) CA model', '(GUZZLER)  (FFS)',
       '(GM-OLDS)  (FFS) CA model', '(GM-OLDS)  (FFS)',
       '(FFS/TRBO) CA model', '(FFS/TRBO)', '(GM-CHEV)  (FFS) CA model',
       '(GM-CHEV)  (FFS)', '(350 V8) (POLICE)   (FFS)',
       '(CALIF)    (FFS,TRBO)', '(POLICE)   (FFS)',
       '(FFS) Fuel Injection', '(GUZZLER)  (TURBO)',
       '(GUZZLER)  (TURBO) CA model', '(ROTARY) CA model', '(ROTARY)',
       '(CALIF)    (DSL,T

Likewise, missing [Engine Descriptors](https://www.fueleconomy.gov/feg/findacarhelp.shtml#engine) can also be replaced with "Unknown":

In [334]:
df['Engine Descriptor'].fillna('Unknown', inplace=True)
nulls = nulls_to_df(df)
nulls.sort_values(by='Nulls', ascending=False)

Nulls  Valids
Drive                               1189   36924
Engine Cylinders                     136   37977
Engine Displacement                  134   37979
Transmission                          11   38102
Fuel Economy Score                     0   38113
...                                  ...     ...
Highway Electricity Consumption        0   38113
Highway Gasoline Consumption (CD)      0   38113
Unrounded Highway MPG (FT2)            0   38113
Highway MPG (FT2)                      0   38113
Manufacturer Code Clean                0   38113

[82 rows x 2 columns]

In [336]:
df['Drive'].unique()

array([nan, '2-Wheel Drive', '4-Wheel or All-Wheel Drive',
       'Rear-Wheel Drive', 'Front-Wheel Drive', '4-Wheel Drive',
       'All-Wheel Drive', 'Part-time 4-Wheel Drive'], dtype=object)

The **Drive** column corresponds to the vehicles drive axle type:
- 2-Wheel Drive
- 4-Wheel Drive*
- 4-Wheel or All-Wheel Drive*
- All-Wheel Drive*
- Front-Wheel Drive
- Part-time 4-Wheel Drive*
- Rear-Wheel Drive
<br>
<br>

*As a caveat, FuelEconomy.gov notes that before "Model Year 2010, EPA did not differentiate between All Wheel Drive and Four Wheel Drive."*
<br>
<br>

Because all vehicles would have this feature, I will replace nulls in this column with "Unknown".

In [337]:
df['Drive'].fillna('Unknown', inplace=True)
nulls = nulls_to_df(df)
nulls.sort_values(by='Nulls', ascending=False)

Nulls  Valids
Engine Cylinders                     136   37977
Engine Displacement                  134   37979
Transmission                          11   38102
Fuel Economy Score                     0   38113
Hatchback Passenger Volume             0   38113
...                                  ...     ...
Highway Gasoline Consumption (CD)      0   38113
Unrounded Highway MPG (FT2)            0   38113
Highway MPG (FT2)                      0   38113
Unrounded Highway MPG (FT1)            0   38113
Manufacturer Code Clean                0   38113

[82 rows x 2 columns]

In [342]:
df['Engine Cylinders'].unique()

array([ 6.,  4.,  8.,  5., 12.,  2.,  3., nan, 10., 16.])

For some strange reason, the values in the **Engine Cylinders** column are encoded as floats even though the numbers represent objects that are countable by natural numbers. Rather than introduce strings in this column for null values, I would prefer to encode these unknowns as 0 so that I can perform numeric analysis if needed. I will also convert all floats to integers.

In [356]:
df['Engine Cylinders'] = df['Engine Cylinders'].fillna(0)
df['Engine Cylinders'] = df['Engine Cylinders'].astype(int)
df['Engine Cylinders'].unique()

array([ 6,  4,  8,  5, 12,  2,  3,  0, 10, 16])

In [357]:
nulls = nulls_to_df(df)
nulls.sort_values(by='Nulls', ascending=False)

Nulls  Valids
Engine Displacement            134   37979
Transmission                    11   38102
Fuel Economy Score               0   38113
Hatchback Passenger Volume       0   38113
4D Luggage Volume                0   38113
...                            ...     ...
Unrounded Highway MPG (FT2)      0   38113
Highway MPG (FT2)                0   38113
Unrounded Highway MPG (FT1)      0   38113
Highway MPG (FT1)                0   38113
Manufacturer Code Clean          0   38113

[82 rows x 2 columns]

In [358]:
df['Engine Displacement'].unique()

array([2.5, 2. , 4.2, 5.3, 1.8, 2.2, 2.1, 5. , 1.5, 2.3, 3. , 2.7, 3.2,
       4.3, 3.8, 5.7, 4.1, 2.8, 6. , 6.2, 1.6, 2.6, 5.2, 5.9, 3.7, 1.4,
       4.9, 5.8, 1.3, 1. , 4.8, 2.4, 1.1, 1.7, 4.7, 6.8, 1.9, 2.9, 3.4,
       nan, 3.9, 5.6, 5.5, 3.3, 3.5, 4. , 1.2, 4.5, 3.6, 3.1, 7.4, 4.6,
       8. , 5.4, 6.5, 4.4, 6.7, 8.3, 6.1, 7. , 8.4, 6.3, 6.6, 6.4, 0.6,
       0. , 0.9])

**Engine Displacement** is "the measure of the cylinder volume swept by all of the pistons of a piston engine, excluding the combustion chambers", according to [Wikipedia](https://en.wikipedia.org/wiki/Engine_displacement). This dataset measures this value in liters. As with the previous column, I will replace null values with 0 so that calculations can be performed.

In [359]:
df['Engine Displacement'].fillna(0, inplace=True)
nulls = nulls_to_df(df)
nulls.sort_values(by='Nulls', ascending=False)

Nulls  Valids
Transmission                    11   38102
Vehicle ID                       0   38113
Fuel Economy Score               0   38113
Hatchback Passenger Volume       0   38113
4D Luggage Volume                0   38113
...                            ...     ...
Unrounded Highway MPG (FT2)      0   38113
Highway MPG (FT2)                0   38113
Unrounded Highway MPG (FT1)      0   38113
Highway MPG (FT1)                0   38113
Manufacturer Code Clean          0   38113

[82 rows x 2 columns]

In [360]:
df['Transmission'].unique()

array(['Manual 5-Speed', 'Automatic 3-Speed', 'Manual 4-Speed',
       'Automatic 4-Speed', 'Manual 3-Speed', 'Manual 4-Speed Doubled',
       nan, 'Automatic (S4)', 'Manual 5 Speed', 'Manual 6-Speed',
       'Automatic (variable gear ratios)', 'Automatic 5-Speed',
       'Automatic (S5)', 'Auto(L4)', 'Auto(L3)', 'Automatic (A1)',
       'Automatic 6-Speed', 'Automatic (S6)', 'Automatic 7-Speed',
       'Automatic (S7)', 'Automatic (S8)', 'Automatic (AV)', 'Auto(AM7)',
       'Automatic (AM5)', 'Auto(AM6)', 'Automatic (AM6)',
       'Automatic (A6)', 'Auto(AV-S7)', 'Manual 7-Speed', 'Auto (AV-S6)',
       'Auto (AV-S8)', 'Automatic (AV-S6)', 'Auto(AV-S6)', 'Auto(AM5)',
       'Auto (AV)', 'Auto(AV-S8)', 'Automatic 8-Speed', 'Auto(AM-S6)',
       'Auto(AM-S7)', 'Automatic 6Speed', 'Automatic 9-Speed',
       'Automatic (S9)', 'Auto(AM-S8)', 'Auto(AM8)', 'Manual(M7)',
       'Auto(AM-S9)', 'Automatic (S10)'], dtype=object)

Null values for **Transmission** can be replaced with "Unknown":

In [361]:
df['Transmission'].fillna('Unknown', inplace=True)
nulls = nulls_to_df(df)
nulls.sort_values(by='Nulls', ascending=False)

Nulls  Valids
Vehicle ID                       0   38113
Hatchback Luggage Volume         0   38113
4D Luggage Volume                0   38113
4D Passenger Volume              0   38113
2D Luggage Volume                0   38113
...                            ...     ...
Highway MPG (FT2)                0   38113
Unrounded Highway MPG (FT1)      0   38113
Highway MPG (FT1)                0   38113
City Utility Factor              0   38113
Manufacturer Code Clean          0   38113

[82 rows x 2 columns]

In [365]:
df[df.isnull().any(axis=1)]

Empty DataFrame
Columns: [Vehicle ID, Year, Make, Model, Class, Drive, Transmission, Transmission Descriptor, Engine Index, Engine Descriptor, Engine Cylinders, Engine Displacement, Turbocharger, Supercharger, Fuel Type, Fuel Type 1, Fuel Type 2, City MPG (FT1), Unrounded City MPG (FT1), City MPG (FT2), Unrounded City MPG (FT2), City Gasoline Consumption (CD), City Electricity Consumption, City Utility Factor, Highway MPG (FT1), Unrounded Highway MPG (FT1), Highway MPG (FT2), Unrounded Highway MPG (FT2), Highway Gasoline Consumption (CD), Highway Electricity Consumption, Highway Utility Factor, Unadjusted City MPG (FT1), Unadjusted Highway MPG (FT1), Unadjusted City MPG (FT2), Unadjusted Highway MPG (FT2), Combined MPG (FT1), Unrounded Combined MPG (FT1), Combined MPG (FT2), Unrounded Combined MPG (FT2), Combined Electricity Consumption, Combined Gasoline Consumption (CD), Combined Utility Factor, Annual Fuel Cost (FT1), Annual Fuel Cost (FT2), Gas Guzzler Tax, Save or Spend (5 Year), Annual Consumption in Barrels (FT1), Annual Consumption in Barrels (FT2), Tailpipe CO2 (FT1), Tailpipe CO2 in Grams/Mile (FT1), Tailpipe CO2 (FT2), Tailpipe CO2 in Grams/Mile (FT2), Fuel Economy Score, GHG Score, GHG Score (Alt Fuel), My MPG Data, 2D Passenger Volume, 2D Luggage Volume, 4D Passenger Volume, 4D Luggage Volume, Hatchback Passenger Volume, Hatchback Luggage Volume, Start Stop Technology, Alternative Fuel/Technology, Electric Motor, Gasoline/Electricity Blended (CD), Vehicle Charger, Alternate Charger, Hours to Charge (120V), Hours to Charge (240V), Hours to Charge (AC 240V), Composite City MPG, Composite Highway MPG, Composite Combined MPG, Range (FT1), City Range (FT1), Highway Range (FT1), Range (FT2), City Range (FT2), Highway Range (FT2), Range (FT2) Clean, Manufacturer Code Clean]
Index: []

[0 rows x 82 columns]

Finally, all columns have had null values removed. However, I still need to inspect the data to ensure that every column has reasonable values.

In [369]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 38113 entries, 0 to 38112
Data columns (total 82 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   Vehicle ID                           38113 non-null  int64  
 1   Year                                 38113 non-null  int64  
 2   Make                                 38113 non-null  object 
 3   Model                                38113 non-null  object 
 4   Class                                38113 non-null  object 
 5   Drive                                38113 non-null  object 
 6   Transmission                         38113 non-null  object 
 7   Transmission Descriptor              38113 non-null  object 
 8   Engine Index                         38113 non-null  int64  
 9   Engine Descriptor                    38113 non-null  object 
 10  Engine Cylinders                     38113 non-null  int32  
 11  Engine Displacement         

After looking through the datatypes of each column, it looks like only **My MPG Data** and **Range (FT2)** have object/string types that should be numbers. Since I already cleaned the data in Range (FT2) and made a new column with numeric data, I can drop this column.
<br>
<br>

Now I need to inspect the contents of the My MPG Data column to see what it contains and whether it should be  modified.

In [372]:
df['My MPG Data'].unique()

array(['N', 'Y'], dtype=object)

After consulting with FuelEconomy.gov as well as inspecting the unique values in this column, I can see that this column simply denotes whether the vehicle has [My MPG Data](https://www.fueleconomy.gov/feg/ws/#ympgVehicle). Therefore, no changes are needed.

In [373]:
df = df.drop('Range (FT2)', axis = 1)

In [374]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 38113 entries, 0 to 38112
Data columns (total 81 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   Vehicle ID                           38113 non-null  int64  
 1   Year                                 38113 non-null  int64  
 2   Make                                 38113 non-null  object 
 3   Model                                38113 non-null  object 
 4   Class                                38113 non-null  object 
 5   Drive                                38113 non-null  object 
 6   Transmission                         38113 non-null  object 
 7   Transmission Descriptor              38113 non-null  object 
 8   Engine Index                         38113 non-null  int64  
 9   Engine Descriptor                    38113 non-null  object 
 10  Engine Cylinders                     38113 non-null  int32  
 11  Engine Displacement         

In [376]:
# df.to_csv('./data/fuel_eco_clean.csv', index = False, header=True, encoding='utf-8-sig')